### PR TITLE
Updates for unit testing docs

### LINKED ISSUE #6344: Unit testing docs
### Contributions

- [X] I have read the contribution docs, and understand what's expected of me.

### Link to the page on docs.getdbt.com requiring updates

https://docs.getdbt.com/docs/build/unit-tests

### What part(s) of the page would you like to see updated?

We got some feedback that suggested the following updates to our documentation:
- Clarify fixture requirements for all dependencies (to avoid null values in output)
- Emphasize satisfying all model logic conditions (i.e. join / where clauses to avoid nulls)
- Under-the-hood code example of how unit tests compile

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by dbeatty10**
## What are you changing in this pull request and why?
Resolves https://github.com/dbt-labs/docs.getdbt.com/issues/6344

## Checklist
- [x] I have reviewed the [Content style guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) so my content adheres to these guidelines.

---
🚀 Deployment available! Here are the direct links to the updated files:

- https://docs-getdbt-com-git-dbeatty-docs-issue-6344-dbt-labs.vercel.app/docs/build/unit-tests

### CODE REVIEW

File: website/docs/docs/build/unit-tests.md [hunk 1]
```diff
@@ -16,7 +16,7 @@ This functionality is only supported in dbt Core v1.8+ or accounts that have opt
...
 
-With dbt Core v1.8 and dbt Cloud environments that have gone versionless by selecting the **Versionless** option, we have introduced an additional type of test to dbt - unit tests. In software programming, unit tests validate small portions of your functional code, and they work much the same way here. Unit tests allow you to validate your SQL modeling logic on a small set of static inputs _before_ you materialize your full model in production. Unit tests enable test-driven development, benefiting developer efficiency and code reliability. 
+With dbt Core v1.8 and dbt Cloud environments that have gone versionless by selecting the **Versionless** option, we have introduced an additional type of test to dbt - unit tests. In software programming, unit tests validate small portions of your functional code, and they work much the same way here. Unit tests allow you to validate your SQL modeling logic on a small set of static inputs _before_ you materialize your full model in production. Unit tests enable test-driven development, benefiting developer efficiency and code reliability.
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md]: Restore original formatting:

```suggestion
With dbt Core v1.8 and dbt Cloud environments that have gone versionless by selecting the **Versionless** option, we have introduced an additional type of test to dbt - unit tests. In software programming, unit tests validate small portions of your functional code, and they work much the same way here. Unit tests allow you to validate your SQL modeling logic on a small set of static inputs _before_ you materialize your full model in production. Unit tests enable test-driven development, benefiting developer efficiency and code reliability. 
```

File: website/docs/docs/build/unit-tests.md [hunk 2]
```diff
@@ -28,7 +28,8 @@ With dbt Core v1.8 and dbt Cloud environments that have gone versionless by sele
 - If your model has multiple versions, by default the unit test will run on *all* versions of your model. Read [unit testing versioned models](/reference/resource-properties/unit-testing-versions) for more information.
 - Unit tests must be defined in a YML file in your `models/` directory.
 - Table names must be [aliased](/docs/build/custom-aliases) in order to unit test `join` logic.
-- Redshift customers need to be aware of a [limitation when building unit tests](/reference/resource-configs/redshift-configs#unit-test-limitations) that requires a workaround. 
+- Redshift customers need to be aware of a [limitation when building unit tests](/reference/resource-configs/redshift-configs#unit-test-limitations) that requires a workaround.
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md]: Restore original formatting:

```suggestion
- Redshift customers need to be aware of a [limitation when building unit tests](/reference/resource-configs/redshift-configs#unit-test-limitations) that requires a workaround. 
```

File: website/docs/docs/build/unit-tests.md [hunk 3]
```diff
@@ -42,7 +43,7 @@ You should unit test a model:
     - `case when` statements when there are many `when`s
     - Truncation
 - When you're writing custom logic to process input data, similar to creating a function.
-- We don't recommend conducting unit testing for functions like `min()` since these functions are tested extensively by the warehouse. If an unexpected issue arises, it's more likely a result of issues in the underlying data rather than the function itself. Therefore, fixture data in the unit test won't provide valuable information.
+- We don't recommend conducting unit testing for functions like `min()` since these functions are tested extensively by the warehouse. If an unexpected issue arises, it's more likely a result of issues in the underlying data rather than the function itself. Therefore, fixture data in the unit test won't provide valuable information.
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md]: Restore original formatting:

```suggestion
- We don't recommend conducting unit testing for functions like `min()` since these functions are tested extensively by the warehouse. If an unexpected issue arises, it's more likely a result of issues in the underlying data rather than the function itself. Therefore, fixture data in the unit test won't provide valuable information.
```

File: website/docs/docs/build/unit-tests.md [hunk 4]
```diff
@@ -81,7 +84,7 @@ check_valid_emails as (
         customers.last_name,
         customers.email,
 	      coalesce (regexp_like(
-            customers.email, '^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$'
+            customers.email, '^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$'
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md]: Restore orignal:

```suggestion
            customers.email, '^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$'
```

File: website/docs/docs/build/unit-tests.md [hunk 5]
```diff
@@ -128,7 +131,7 @@ unit_tests:
 
 The previous example defines the mock data using the inline `dict` format, but you can also use `csv` or `sql` either inline or in a separate fixture file. Store your fixture files in a `fixtures` subdirectory in any of your [test paths](/reference/project-configs/test-paths). For example, `tests/fixtures/my_unit_test_fixture.sql`. 
 
-When using the `dict` or `csv` format, you only have to define the mock data for the columns relevant to you. This enables you to write succinct and _specific_ unit tests.
+When using the `dict` or `csv` format, you only have to define the mock data for the columns relevant to you. This enables you to write succinct and _specific_ unit tests.
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md]: Restore original:

```suggestion
When using the `dict` or `csv` format, you only have to define the mock data for the columns relevant to you. This enables you to write succinct and _specific_ unit tests.
```

File: website/docs/docs/build/unit-tests.md [hunk 6]
```diff
@@ -190,9 +192,9 @@ actual differs from expected:
 
 ```
 
-The clever regex statement wasn’t as clever as initially thought, as the model incorrectly flagged `cool@example.com` as an invalid email address.
+The clever regex statement wasn’t as clever as initially thought, as the model incorrectly flagged `cool@example.com` as an invalid email address.
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md]: Restore original:

```suggestion
The clever regex statement wasn’t as clever as initially thought, as the model incorrectly flagged `cool@example.com` as an invalid email address.
```

File: website/docs/docs/build/unit-tests.md [hunk 7]
```diff
@@ -190,9 +192,9 @@ actual differs from expected:
...
 
-Updating the regex logic to `'^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$'` (those pesky escape characters) and rerunning the unit test solves the problem:
+Updating the regex logic to `'^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$'` (those pesky escape characters) and rerunning the unit test solves the problem:
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md]: Restore original:

```suggestion
Updating the regex logic to `'^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$'` (those pesky escape characters) and rerunning the unit test solves the problem:
```

File: website/docs/docs/build/unit-tests.md [hunk 8]
```diff
@@ -29,6 +29,7 @@ With dbt Core v1.8 and dbt Cloud environments that have gone versionless by sele
 - Unit tests must be defined in a YML file in your `models/` directory.
 - Table names must be [aliased](/docs/build/custom-aliases) in order to unit test `join` logic.
 - Redshift customers need to be aware of a [limitation when building unit tests](/reference/resource-configs/redshift-configs#unit-test-limitations) that requires a workaround. 
+- All references (`ref()`) used in your model must be included in the unit test configuration as input fixtures, even if they do not directly affect the logic being tested. If these references are missing, you may encounter "node not found" errors during compilation.
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md:32]: Should be [`ref`](https://docs.getdbt.com/reference/dbt-jinja-functions/ref) _or_ [`source`](https://docs.getdbt.com/reference/dbt-jinja-functions/source).

See https://github.com/dbt-labs/docs.getdbt.com/issues/6331 for a related docs issue asking for a unified super concept that describes both `ref` and `source` calls.

File: website/docs/docs/build/unit-tests.md [hunk 9]
```diff
@@ -56,6 +57,8 @@ Use the [resource type](/reference/global-configs/resource-type) flag `--exclude
 
 ## Unit testing a model
 
+When defining mock data for a unit test, it's crucial to include all necessary input values that satisfy the entire model logic. This means including values that fulfill any `WHERE` clauses, `JOIN` conditions, or other constraints present in the model, even if they do not seem directly related to the specific logic being tested. Failing to do so may lead to errors or unexpected null values in the unit test results.
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md:60]: We probably don't want to lead with this paragraph. Instead, we'd probably want to move it somewhere else. Possibly under a new "pitfalls" section?

File: website/docs/docs/build/unit-tests.md [hunk 10]
```diff
@@ -319,6 +322,36 @@ Exit codes differ from data test success and failure outputs because they don't
 Learn about [exit codes](/reference/exit-codes) for more information.
 
 
+### Common Pitfalls
+> - **Missing Fixtures for Referenced Models**: When creating a unit test, all referenced models must be declared as mock inputs. Missing any referenced model, even if it isn't directly involved in the specific logic being tested, will lead to compilation errors such as "node not found."
+> - **Not Satisfying `WHERE` or `JOIN` Logic**: Ensure that the mock data meets all conditions in the model, such as `WHERE` clauses or `JOIN` requirements. If these conditions are not met, the unit test will either return null rows or fail to execute properly. This often involves adding rows for auxiliary data tables, like locations or transactions, to satisfy joins and filters.
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md:327]: We'd probably want to give a more detailed example here. Otherwise, it's hard to determine what exactly we mean here. Maybe it would be best to create a post in Discourse with all the details and then link to it?

File: website/docs/docs/build/unit-tests.md [hunk 11]
```diff
@@ -319,6 +322,36 @@ Exit codes differ from data test success and failure outputs because they don't
...
+>    transaction_id,
+>    sum(cash_value) as donated_cash
+>  from __dbt__cte__stg_donations
+>  group by transaction_id
+> )
+> select * from __dbt__cte__stg_transactions;
+> ```
+> Understanding this process will help ensure you configure your unit tests correctly and avoid common issues.
+
+
```
Comment (dbeatty10) [inline website/docs/docs/build/unit-tests.md:354]: This is interesting implementation detail that helps in troubleshooting when there is some kind of error. In general, this kind of detail doesn't feel like it should go in product docs though because the maintainers are free to change the details at will. Maybe we move it to Markdown docs within the dbt-core or dbt-adapters repo instead?

**Review by dbeatty10 — COMMENTED** [review]
Left some comments on updates to make in what we say and where we say it.

**Comment by jairus-m** [comment]
Hey @dbeatty10!

I noticed this draft and wanted to point you to #6330 .

I hit super similar updates and give an example! hope this may be helpful whether it's approved or not


Here is that proposed section:

## Caveats to unit test configuration
As you start to unit test logic in models that contain complex joins or conditions, there are a few things to consider:

All references in your tested model must be declared in the configuration regardless if they're directly relevant to the core unit test logic or not.

In addition, when mocking data for your unit test, ensure it satisfies all conditions in your model's SQL, including those that may not be directly related to the specific logic that you're testing. 

This includes:
- WHERE clause conditions
- JOIN conditions
- Any other logic that affects row selection or transformations

Continuing the example above, we can create an updated version of the customers table called `dim_customers_with_orders`. Here, we join an additional orders table and add a condition in the WHERE clause:
<file name='dim_customers_with_transactions.sql'>

```sql
with customers as (

    select * from {{ ref('stg_customers') }}

),

-- Addition of a ref() that is not used in core unit test
orders as (

  select
      customer_id,
      count(distinct order_id) as num_orders,
      sum(order_total) as total_order_value
  from {{ ref('stg_orders') }}
  group by customer_id

),

accepted_email_domains as (

    select * from {{ ref('top_level_email_domains') }}

)

select
    customers.customer_id,
    customers.first_name,
    customers.last_name,
    customers.email,
    coalesce(regexp_like(
		customers.email, '^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$'
		)
	= true
	and accepted_email_domains.tld is not null,
	false) as is_valid_email_address
    orders.num_orders,
    orders.total_order_value
from customers
left join orders
    on customers.customer_id = orders.customer_id
left join accepted_email_domains
    on customers.email_top_level_domain = lower(accepted_email_domains.tld)
-- Addition of WHERE condition
where customers.age is >= 21
```
</file>

With the added CTE (orders) and the additional statement in the WHERE clause, we now need to add mock data that satifies `where age >= 21` as well as add an empty input reference to `stg_orders`. The resulting unit test configuration will now be:

<file name='unit_test.yml'> 

```yaml
unit_tests:
  - name: test_is_valid_email_address
    description: "Check my is_valid_email_address logic captures all known edge cases - emails without ., emails without @, and emails from invalid domains."
    model: dim_customers
    given:
      - input: ref('stg_customers')
        rows:
          - {email: cool@example.com,    email_top_level_domain: example.com,  age: 21}
          - {email: cool@unknown.com,    email_top_level_domain: unknown.com,  age: 21}
          - {email: badgmail.com,        email_top_level_domain: gmail.com,    age: 21}
          - {email: missingdot@gmailcom, email_top_level_domain: gmail.com,    age: 21}
      - input: ref('top_level_email_domains')
        rows:
          - {tld: example.com}
          - {tld: gmail.com}
      - input: ref('stg_orders')
        rows:
          - {}
    expect:
      rows:
        - {email: cool@example.com,    is_valid_email_address: true}
        - {email: cool@unknown.com,    is_valid_email_address: false}
        - {email: badgmail.com,        is_valid_email_address: false}
        - {email: missingdot@gmailcom, is_valid_email_address: false}

```
</file>

While the unit test in this example was for a specific column, it’s important to remember that mock inputs are going to be passed through the entire model as a whole and therefore, must be declared as such and configured properly to meet all the model’s expectations.

**Comment by runleonarun** [comment]
Hey @dbeatty10! In an effort to clear our Pull requests backlog, I'm going to close this but added an `icebox` label. It's also connected to the issue for future inspo.
